src/optlang/tests/test_osqp_interface.py

```diff
@@ -0,0 +1,627 @@
+# Copyright (c) 2013 Novo Nordisk Foundation Center for Biosustainability, DTU.
+# See LICENSE for details.
+
+import copy
+import os
+import pickle
+import random
+import unittest
+
+try:  # noqa: C901
+    import osqp
```

Comment:
Tested this with `cuosqp` and it does not work for some of the classes of tests (as we would expect). Not that it matters for CI, but would it be fine to have some tests that should pass for `cuosqp` in case it is installed?

Reply:
Yeah, it fixes some of the parameters like lambda. What tests fail? I would suspect the LP problems since those have a lambda adjustment in the code that is probably ignored. I'll add a warning when using cuosqp.

Reply:
The tests that fail with `cuosqp` are those under `ConstraintTestCase` and `ModelTestCase.test_binary_variables`. There could be more. The thing is that the tests don't fail, they error out:
```
test_binary_variables (optlang.tests.test_osqp_interface.ModelTestCase) ... /home/georg/.virtualenvs/opt/lib/python3.9/site-packages/cuosqp/utils.py:20: UserWarning: Linear system solver not recognized. Using default solver CUDA PCG.
  warn("Linear system solver not recognized. " +
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for failed: cudaErrorNoKernelImageForDevice: no kernel image is available for execution on the device
[1]    19842 abort (core dumped)  python -m nose src/optlang/tests -v
```
That's why I thought that they don't have a linear solver. Does it work for you? It might be a problem with my cuda version (using 11.2), but they say v11 is supported (other libraries which use cuda work for me and some of the OSQP tests pass).

Reply:
This looks more like an issue with the compute capability of the used card. I think I tried it on Google Colab before and it kind of worked.

Reply:
I could reproduced the error on colab https://gist.github.com/carrascomj/a4d91a926814646e36592a853516c7ac
Is there something that I am not installing maybe?

Reply:
I think you just did not change the Runtime to one with a CPU. If you do that it does run but it is slow due to the setting issue and some tests fail because it can't hit the needed accuracy for LPs.

Reply:
RIght! I had changed but when I restarted the runtime it went back to CPU for some reason. It works (with some caveats as you said), so this PR looks done to me.

Reply:
I will add a warning ng message for cuosqp on load to make clear that is an experimental feature. Since you have already worked through it, would be great if you could add a review to the PR, thanks!

Reply:
I don't have write access either (although you may have it now?). First thing on Monday, I will ask Niko to approve it so it can be finally merged. Also, please remember the docs! Something along the lines of
> OSQP: `osqp` | `cuosqp`
> OSQP is an efficient open source solver for Quadratic Programs. It is self-contained and can be installed via pip. Alternatively, cuOSQP provides a cuda-enabled implementation.

under `installation.rst` would suffice, I guess.

Reply:
OK, I have write access now so feel free to merge it.